nextstrain/cli/remote/nextstrain_dot_org.py

```diff
@@ -0,0 +1,697 @@
+                            # which is itself a dataset that exists, e.g.
+                            # `nextstrain remote download -r a/b` when the
+                            # following datasets exist:
+                            #
+                            #       a/b
+                            #       a/b/x
+                            #       a/b/y
+                            #       a/b/z
+                            #
+                            # then we need to include a little of the given
+                            # path in the local filenames, e.g. strip only "a",
+                            # leaving "b".  Otherwise, e.g. if a/b doesn't
+                            # exist itself only a/b/x, etc, then we can strip
+                            # "a/b" entirely.
+                            if path in {r.path for r in resources}:
+                                base_path = path.parent
+                            else:
+                                base_path = path
+
+                            local_name = (
+                                str(resource.path.relative_to(base_path))
+                                    .lstrip("/")
+                                    .replace("/", "_"))
+                        else:
+                            local_name = resource.path.name
+
+                        destination = local_path / local_name
+                    else:
+                        destination = local_path
+
+                    if not subresource.primary:
+                        destination = destination.with_name(f"{destination.with_suffix('').name}_{sidecar_suffix(subresource.media_type)}")
+
+                    destination = destination.with_suffix(subresource.file_extension)
+
+                    yield source, destination
+
+                    # Stream response data to local file
+                    with destination.open("w") as local_file:
+                        for chunk in response.iter_content(chunk_size = None, decode_unicode = True):
+                            local_file.write(chunk)
+
+
+def ls(url: urllib.parse.ParseResult) -> Iterable[str]:
+    """
+    List the datasets and narratives deployed at the given nextstrain.org *url*.
+    """
+    path = remote_path(url)
+
+    return [
+        api_endpoint(item.path)
+            for item
+             in _ls(path, recursively = True)
+    ]
+
+
+def _ls(path: NormalizedPath, recursively: bool = False, http: requests.Session = None, authorization: dict = None):
+    """
+    List the :class:`Resource`(s) available at *path*.
+
+    If *recursively* is false (the default), then only an exact match of
+    *path* is returned, if any.  If *recursively* is true, then all resources
+    at or beneath *path* are returned.
+
+    If *http* is not provided, a new :class:`requests.Session` will be created.
+
+    If *authorization* is not provided, :func:`auth_headers` will be called to
+    obtain it.
```

Comment:
Is there a specific docstring format being followed here? Looks like restructured text but I don't see this being rendered on any page in the preview.

Reply:
rST, yes, but the docstrings are somewhat loose and vary with when they were written. The internal interfaces aren't included in the Sphinx docs, as I'm not really sure who the audience would be for that. Augur [includes its internal interfaces](https://docs.nextstrain.org/projects/augur/en/stable/api/api.html) (generated from docstrings), but it actually has an expectation of being used as a library unlike the Nextstrain CLI.